index/upside_down/reader.go

```diff
@@ -104,11 +105,23 @@ func (r *UpsideDownCouchTermFieldReader) Next(preAlloced *index.TermFieldDoc) (*
 
 func (r *UpsideDownCouchTermFieldReader) Advance(docID index.IndexInternalID, preAlloced *index.TermFieldDoc) (*index.TermFieldDoc, error) {
 	if r.iterator != nil {
-		tfr := NewTermFrequencyRow(r.term, r.field, docID, 0, 0)
-		r.iterator.Seek(tfr.Key())
+		if r.tfrNext == nil {
+			r.tfrNext = &TermFrequencyRow{}
+		}
+		tfr := InitTermFrequencyRow(r.tfrNext, r.term, r.field, docID, 0, 0)
+		keySize := tfr.KeySize()
+		if cap(r.keyBuf) < keySize {
+			r.keyBuf = make([]byte, keySize)
+		}
```

Comment:
Yah, pretty messy now.  I'll throw up a separate, future PR for something like that.